File: android/sentinel-core/src/main/kotlin/com/sentinel/sdk/security/PayloadSigner.kt
package com.sentinel.sdk.security

import com.sentinel.sdk.models.SentinelSecurityData
import java.security.MessageDigest
import java.util.UUID
import javax.crypto.Mac
import javax.crypto.spec.SecretKeySpec
import kotlin.io.encoding.Base64
import kotlin.io.encoding.ExperimentalEncodingApi

class PayloadSigner(private val secretKey: String? = null) {

    fun sign(
        payloadJson: String,
        nonce: String = UUID.randomUUID().toString(),
        tamperDetected: Boolean = false
    ): SentinelSecurityData {
        val payloadHash = sha256Hex(payloadJson.toByteArray(Charsets.UTF_8))
        val signature = generateSignature(payloadHash, nonce)

        return SentinelSecurityData(
            payloadHash = payloadHash,
            signature = signature,
            nonce = nonce,
            tamperDetected = tamperDetected
        )
    }

    fun sha256Hex(bytes: ByteArray): String {
        val digest = MessageDigest.getInstance("SHA-256").digest(bytes)
        return digest.joinToString("") { "%02x".format(it) }
    }

    @OptIn(ExperimentalEncodingApi::class)
    fun generateSignature(payloadHash: String, nonce: String): String {
        val message = "$payloadHash:$nonce"
        val messageBytes = message.toByteArray(Charsets.UTF_8)

        return if (!secretKey.isNullOrEmpty()) {
            val keySpec = SecretKeySpec(secretKey.toByteArray(Charsets.UTF_8), "HmacSHA256")
            val mac = Mac.getInstance("HmacSHA256")
            mac.init(keySpec)
            val hmacBytes = mac.doFinal(messageBytes)
            Base64.encode(hmacBytes)
        } else {
            val digest = MessageDigest.getInstance("SHA-256").digest(messageBytes)
            Base64.encode(digest)
        }
    }

    fun verify(payloadHash: String, nonce: String, signature: String): Boolean {
        val expected = generateSignature(payloadHash, nonce)
        val expectedBytes = expected.toByteArray(Charsets.UTF_8)
        val signatureBytes = signature.toByteArray(Charsets.UTF_8)
        return MessageDigest.isEqual(expectedBytes, signatureBytes)
    }
}

File: ios/Sources/SentinelSDK/Security/PayloadSigner.swift
import Foundation
#if canImport(CryptoKit)
import CryptoKit
#endif

/// Generates cryptographic checksums and signatures for captured payloads.
public final class PayloadSigner {

    private let secretKey: String?

    public init(secretKey: String? = nil) {
        self.secretKey = secretKey
    }

    public func sign(data: Data, nonce: String = UUID().uuidString) -> SentinelSecurityData {
        let hashString = sha256Hex(data: data)
        let signature = generateSignature(payloadHash: hashString, nonce: nonce)

        return SentinelSecurityData(
            payloadHash: hashString,
            signature: signature,
            nonce: nonce,
            tamperDetected: false
        )
    }

    public func sha256Hex(data: Data) -> String {
        #if canImport(CryptoKit)
        let digest = SHA256.hash(data: data)
        return digest.map { String(format: "%02hhx", $0) }.joined()
        #else
        return "\(data.count)"
        #endif
    }

    public func generateSignature(payloadHash: String, nonce: String) -> String {
        let message = "\(payloadHash):\(nonce)"
        guard let messageData = message.data(using: .utf8) else { return "" }

        #if canImport(CryptoKit)
        if let key = secretKey, !key.isEmpty, let keyData = key.data(using: .utf8) {
            let symmetricKey = SymmetricKey(data: keyData)
            let mac = HMAC<SHA256>.authenticationCode(for: messageData, using: symmetricKey)
            return Data(mac).base64EncodedString()
        } else {
            let digest = SHA256.hash(data: messageData)
            return Data(digest).base64EncodedString()
        }
        #else
        return messageData.base64EncodedString()
        #endif
    }

    public func verify(payloadHash: String, nonce: String, signature: String) -> Bool {
        let expected = generateSignature(payloadHash: payloadHash, nonce: nonce)
        return expected == signature
    }
}
